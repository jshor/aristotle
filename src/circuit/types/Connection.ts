import CircuitNode from '../base/CircuitNode'

class Connection {
  node: CircuitNode
  id: string

  constructor (node: CircuitNode, id: string) {
    this.node = node
    this.id = id
  }
}

export default Connection
