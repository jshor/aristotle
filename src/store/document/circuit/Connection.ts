import CircuitNode from './CircuitNode'

class Connection {
  node: CircuitNode
  id: string

  constructor (node: CircuitNode, id: string) {
    this.node = node
    this.id = id
  }
}

export default Connection
