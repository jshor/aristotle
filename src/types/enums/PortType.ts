enum PortType {
  Output = 0,
  Input
}

export default PortType
