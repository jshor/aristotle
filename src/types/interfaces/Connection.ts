import BaseItem from './BaseItem'

interface Connection extends BaseItem {
  /** ID of the connected source port. */
  source: string // TODO: rename to sourceId?
  /** ID of the connected target port. */
  target: string // TODO: rename to targetId?
  /** ID of the connection chain (a linked series of connections joined by freeports). */
  connectionChainId: string
  /** ID of the group that this connection belongs to. */
  groupId: string | null
}

export default Connection
