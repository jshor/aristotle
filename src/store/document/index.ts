import { StoreGetters, defineStore } from 'pinia'
import { DocumentState, state } from './state'
import * as clipboardActions from './actions/clipboard'
import * as connectionActions from './actions/connection'
import * as editorActions from './actions/editor'
import * as groupingActions from './actions/grouping'
import * as itemActions from './actions/item'
import * as panningActions from './actions/panning'
import * as portActions from './actions/port'
import * as positioningActions from './actions/positioning'
import * as oscillatorActions from './actions/oscillator'
import * as rotationActions from './actions/rotation'
import * as selectionActions from './actions/selection'
import * as simulationActions from './actions/simulation'
import * as sizingActions from './actions/sizing'
import * as snappingActions from './actions/snapping'
import * as undoRedoActions from './actions/undoRedo'
import * as zIndexActions from './actions/zIndex'
import * as zoomingActions from './actions/zooming'
import * as getters from './getters'
import sortByZIndex from '@/utils/sortByZIndex'
import BaseItem from '@/types/interfaces/BaseItem'
import Item from '@/types/interfaces/Item'
import Connection from '@/types/interfaces/Connection'

export const createDocumentStore = (documentId: string) => defineStore({
  id: documentId,
  state,
  getters: {
    ...getters,

    baseItems (state) {
      return (Object
        .values(state.connections) as BaseItem[])
        .concat(Object.values(state.items) as BaseItem[])
        .sort(sortByZIndex)
    },

    hasSelection (state) {
      return state.selectedItemIds.size > 0 || state.selectedConnectionIds.size > 0
    },

    hasSelectedItems (state) {
      return state.selectedItemIds.size > 0
    },

    canGroup (state) {
      let ungroupedCount = 0
      let groupedCount = 0

      const hasUngrouped = (list: Set<string>, map: Record<string, Connection | Item>) => {
        for (const id of list) {
          map[id].groupId
            ? groupedCount++
            : ungroupedCount++

          if (ungroupedCount > 1 || (groupedCount && ungroupedCount)) {
            // at least two items are ungrouped, or at least one item is grouped and at least one item is ungrouped
            return true
          }
        }

        return false
      }

      return hasUngrouped(state.selectedItemIds, state.items) || hasUngrouped(state.selectedConnectionIds, state.connections)
    },

    canUngroup (state) {
      return state.selectedGroupIds.size > 0
    },

    canUndo (state) {
      return state.undoStack.length > 0
    },

    canRedo (state) {
      return state.redoStack.length > 0
    }
  },
  actions: {
    ...clipboardActions,
    ...connectionActions,
    ...editorActions,
    ...groupingActions,
    ...itemActions,
    ...panningActions,
    ...portActions,
    ...positioningActions,
    ...oscillatorActions,
    ...rotationActions,
    ...selectionActions,
    ...simulationActions,
    ...snappingActions,
    ...sizingActions,
    ...undoRedoActions,
    ...zIndexActions,
    ...zoomingActions
  }
})

export type DocumentStore = ReturnType<typeof createDocumentStore>

export type DocumentStoreInstance = ReturnType<DocumentStore>
