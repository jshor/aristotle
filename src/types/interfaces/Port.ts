import BinaryWavePulse from '@/store/document/oscillator/BinaryWavePulse'
import PortType from '../enums/PortType'
import Direction from '../enums/Direction'
import Point from './Point'

interface Port {
  id: string
  name: string
  elementId: string
  virtualElementId?: string
  position: Point
  type: PortType
  rotation: number
  orientation: Direction
  value: number
  hue: number
  isFreeport: boolean
  isMonitored: boolean
  connectedPortIds: string[]
  wave?: BinaryWavePulse
}

export default Port
