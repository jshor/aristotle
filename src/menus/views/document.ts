import { MenuFactory } from '@/types/interfaces/MenuFactory'
import { DocumentStore } from '@/store/document'
import { createHelpSubmenu } from '../submenus/help'
import { createEditSubmenu } from '../submenus/edit'
import { createFileSubmenu } from '../submenus/file'
import { createViewSubmenu } from '../submenus/view'
import { t } from '@/utils/i18n'

/**
 * Creates the document application menu.
 */
export const createDocumentMenu: MenuFactory = (useDocumentStore?: DocumentStore) => {
  return [
    {
      label: t('menu.file.parent'),
      submenu: createFileSubmenu(useDocumentStore)
    },
    {
      label: t('menu.edit.parent'),
      submenu: createEditSubmenu(useDocumentStore)
    },
    {
      label: t('menu.view.parent'),
      submenu: createViewSubmenu(useDocumentStore)
    },
    {
      label: t('menu.help.parent'),
      role: 'help',
      submenu: createHelpSubmenu()
    }
  ]
}
