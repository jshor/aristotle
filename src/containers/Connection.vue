<template>
  <draggable
    :style="{ zIndex }"
    :position="position"
    :is-selected="isSelected"
    :allow-touch-drag="freeportId !== null"
    @touchhold="onDragStart"
    @drag="onDrag"
    @drag-end="onDragEnd"
    @select="k => selectItem(id, k)"
    @deselect="deselectItem(id)"
    @contextmenu="onContextMenu"
  >
    <wire
      :source="source"
      :target="target"
      :geometry="geometry"
      :is-selected="isSelected"
      :is-preview="isPreview"
      :flash="flash"
      :label="'TODO'"
      ref="root"
      data-test="wire"
    />
  </draggable>
</template>

<script lang="ts">
import { v4 as uuid } from 'uuid'
import { ComponentPublicInstance, defineComponent, PropType, ref, computed } from 'vue'
import Draggable from '@/components/interactive/Draggable.vue'
import Wire from '@/components/Wire.vue'
import { DocumentStore } from '@/store/document'
import boundaries from '@/store/document/geometry/boundaries'
import renderLayout from '@/store/document/geometry/wire'
import editorContextMenu from '@/menus/context/editor'

export default defineComponent({
  name: 'Connection',
  components: {
    Draggable,
    Wire
  },
  props: {
    /** Whether or not this connection is selected. */
    isSelected: {
      type: Boolean,
      default: false
    },

    /** ID of the connection. */
    id: {
      type: String,
      required: true
    },

    /** CSS z-index value. */
    zIndex: {
      type: Number,
      default: 0
    },

    /** Whether or not this connection is just for previewing. */
    isPreview: {
      type: Boolean,
      default: false
    },

    /** Whether or not the item should show a flash once to the user. */
    flash: {
      type: Boolean,
      default: false
    },

    /** Document store instance. */
    store: {
      type: Function as PropType<DocumentStore>,
      required: true
    }
  },
  setup (props) {
    const store = props.store()
    const connection = ref(store.connections[props.id])
    const root = ref<ComponentPublicInstance<HTMLElement>>()
    const source = computed(() => store.ports[store.connections[props.id].source])
    const target = computed(() => store.ports[store.connections[props.id].target])
    const geometry = computed(() => renderLayout(source.value, target.value))
    const position = computed(() => {
      const topLeft = boundaries.getExtremePoint('min', source.value.position, target.value.position)

      return {
        y: topLeft.y + geometry.value.minY,
        x: topLeft.x + geometry.value.minX
      }
    })

    const freeportId = ref<string | null>(null)

    /**
     * Mouse button down event handler.
     *
     * This will inform the component that the mouse is down and ready to create a new freeport, if it moves.
     */
    function onDragStart ({ touches }: TouchEvent) {
      // TODO: bug: if the cursor drags only 1px (or some insufficient amount to create a new freeport) then the connection will inadvertantly be deleted
      createFreeport({
        x: touches[0].clientX,
        y: touches[0].clientY
      })
    }

    function createFreeport (position: Point) {
      if (!root.value?.$el || connection.value.groupId !== null) return

      // directly change the DOM style (as opposed to using v-show/refs) to avoid "flickering" while VDOM updates
      root.value.$el.style.opacity = '0'
      freeportId.value = uuid()
      navigator.vibrate(20)

      store.createFreeport({
        itemId: freeportId.value,
        inputPortId: uuid(),
        outputPortId: uuid(),
        connectionChainId: connection.value.connectionChainId,
        sourceId: source.value.id,
        targetId: target.value.id,
        value: source.value.value
      })

      store.setSnapBoundaries(freeportId.value)
      store.dragItem(freeportId.value, position)
    }

    /**
     * Mouse move event handler.
     *
     * If the mouse is held down on the wire, then movement in both the x and y directions
     * will cause a new freeport to be created in that position.
     *
     * This will only create a new freeport once per mousedown-move cycle.
     */
    async function onDrag (position: Point) {
      if (!freeportId.value) {
        createFreeport(position)
      } else {
        store.dragItem(freeportId.value, position)
      }
    }

    /**
     * Mouse up event handler.
     */
    function onDragEnd () {
      if (freeportId) {
        // after dragging has finished, two new connections are established
        // this connection is now ready to destroy itself
        store.destroyConnection({
          source: connection.value.source,
          target: connection.value.target
        })
      }

      freeportId.value = null
    }

    function onContextMenu () {
      if (!props.isSelected) {
        store.selectItem(props.id)
      }

      window.api.showContextMenu(editorContextMenu(props.store))
    }

    return {
      root,
      geometry,
      position,
      onDragStart,
      onDrag,
      onDragEnd,
      onContextMenu,
      source,
      target,
      freeportId,
      selectItem: store.selectItem,
      deselectItem: store.deselectItem
    }
  }
})
</script>
