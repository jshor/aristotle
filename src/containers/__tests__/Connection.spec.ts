import { mount, VueWrapper } from '@vue/test-utils'
import { setActivePinia, createPinia } from 'pinia'
import ConnectionComponent from '../Connection.vue'
import { createDocumentStore, DocumentStore, DocumentStoreInstance } from '@/store/document'
import { createConnection, createControlPoint, createPort, stubAll } from '@/store/document/actions/__tests__/__helpers__'
import PortType from '@/types/enums/PortType'
import Port from '@/types/interfaces/Port'
import Connection from '@/types/interfaces/Connection'

setActivePinia(createPinia())

describe('Item container', () => {
  let wrapper: VueWrapper
  let storeDefinition: DocumentStore
  let store: DocumentStoreInstance
  let source: Port
  let target: Port
  let connection: Connection

  const connectionId = 'connection'
  const props = {
    zIndex: 1,
    isSelected: true,
    id: connectionId,
    flash: false
  }

  beforeEach(() => {
    storeDefinition = createDocumentStore('test-id')
    store = storeDefinition()

    source = createPort('source', 'item', PortType.Input, { name: 'Source Port' })
    target = createPort('target', 'item', PortType.Input, { name: 'Target Port' })
    connection = createConnection(connectionId, source.id, target.id)

    store.connections = { connection }
    store.ports = { source, target }

    wrapper = mount(ConnectionComponent, {
      props: {
        ...props,
        store: storeDefinition
      }
    })
  })

  afterEach(() => {
    vi.resetAllMocks()
  })

  it('should match the snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('when the centroid is focused', () => {
    beforeEach(async () => {
      stubAll(store, [
        'deselectAll',
        'setConnectionSelectionState'
      ])

      await wrapper
        .find('[data-test="centroid"]')
        .trigger('focus')
    })

    it('should select the connection', () => {
      expect(store.setConnectionSelectionState).toHaveBeenCalledTimes(1)
      expect(store.setConnectionSelectionState).toHaveBeenCalledWith(connectionId, true)
    })

    it('should show the bounding box', () => {
      expect(wrapper.find('[data-test="bounding-box"]').exists()).toBe(true)
    })

    it('should deselect all other items', () => {
      expect(store.deselectAll).toHaveBeenCalledTimes(1)
    })

    it('should hide the bounding box when the centroid is blurred', async () => {
      await wrapper
        .find('[data-test="centroid"]')
        .trigger('blur')

      expect(wrapper.find('[data-test="bounding-box"]').exists()).toBe(false)
    })
  })

  describe('when dragging a wire', () => {
    const position = { x: 10, y: 20 }
    const index = 0

    beforeEach(() => {
      stubAll(store, [
        'cacheState',
        'deselectAll',
        'setControlPointSnapBoundaries',
        'setSelectionPosition',
        'addControlPoint',
        'dragControlPoint'
      ])
    })

    it('should move the new control point to the emitted position', async () => {
      await wrapper
        .findComponent({ name: 'WireDraggable' })
        .vm
        .$emit('dragStart', position)
      await wrapper
        .findComponent({ name: 'WireDraggable' })
        .vm
        .$emit('move', position)

      expect(store.dragControlPoint).toHaveBeenCalledTimes(1)
      expect(store.dragControlPoint).toHaveBeenCalledWith(connectionId, position, index)
    })

    it('snap', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('should move the entire selection when it is part of a group', async () => {
      const startPosition = { x: 1, y: 5 }
      const movePosition = { x: 10, y: 20 }

      store.connections[connectionId].groupId = 'group-id'
      store.zoomLevel = 1.2

      await wrapper
        .findComponent({ name: 'WireDraggable' })
        .vm
        .$emit('move', startPosition)

      expect(store.setSelectionPosition).toHaveBeenCalledTimes(1)
      expect(store.setSelectionPosition).toHaveBeenCalledWith({
        x: startPosition.x / store.zoomLevel,
        y: startPosition.y / store.zoomLevel
      })

      await wrapper.vm.$nextTick()
      await wrapper
        .findComponent({ name: 'WireDraggable' })
        .vm
        .$emit('move', movePosition)

      expect(store.setSelectionPosition).toHaveBeenCalledTimes(2)
      expect(store.setSelectionPosition).toHaveBeenCalledWith({
        x: (movePosition.x - startPosition.x) / store.zoomLevel,
        y: (movePosition.y - startPosition.y) / store.zoomLevel
      })
    })
  })

  describe('when dragging a control point', () => {
    const position = { x: 10, y: 20 }
    const index = 0

    beforeEach(() => {
      stubAll(store, [
        'cacheState',
        'commitCachedState',
        'deselectAll',
        'setControlPointSnapBoundaries',
        'setSelectionPosition',
        'addControlPoint',
        'dragControlPoint'
      ])

      store.connections[connectionId].controlPoints = [createControlPoint(), createControlPoint()]
    })

    it('should move the new control point to the emitted position', async () => {
      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('dragStart', position)

      expect(store.cacheState).toHaveBeenCalledTimes(1)
      expect(store.setControlPointSnapBoundaries).toHaveBeenCalledTimes(1)
      expect(store.setControlPointSnapBoundaries).toHaveBeenCalledWith(connectionId, expect.any(Array), index)

      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('drag', position)

      expect(store.dragControlPoint).toHaveBeenCalledTimes(1)
      expect(store.dragControlPoint).toHaveBeenCalledWith(connectionId, position, index)

      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('dragEnd', position)

      expect(store.commitCachedState).toHaveBeenCalledTimes(1)
    })

    it('should move the entire selection when the connection is part of a group', async () => {
      const startPosition = { x: 1, y: 5 }
      const movePosition = { x: 10, y: 20 }

      store.connections[connectionId].groupId = 'group-id'
      store.zoomLevel = 1.2

      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('dragStart', startPosition)

      expect(store.cacheState).toHaveBeenCalledTimes(1)

      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('drag', movePosition)

      expect(store.setSelectionPosition).toHaveBeenCalledTimes(1)
      expect(store.setSelectionPosition).toHaveBeenCalledWith({
        x: (movePosition.x - startPosition.x) / store.zoomLevel,
        y: (movePosition.y - startPosition.y) / store.zoomLevel
      })

      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('dragEnd', position)

      expect(store.commitCachedState).toHaveBeenCalledTimes(1)
    })
  })

  describe('context menus', () => {
    describe('when the connection context menu is requested', () => {
      beforeEach(() => {
        stubAll(store, [
          'setConnectionSelectionState'
        ])
      })

      it('should select the connection if it is not already selected', async () => {
        await wrapper.setProps({ isSelected: false })
        await wrapper
          .findComponent({ name: 'WireDraggable' })
          .trigger('contextmenu')

        expect(store.setConnectionSelectionState).toHaveBeenCalledTimes(1)
        expect(store.setConnectionSelectionState).toHaveBeenCalledWith(connectionId, true)
      })

      it('should not select the connection if it is already selected', async () => {
        await wrapper.setProps({ isSelected: true })
        await wrapper
          .findComponent({ name: 'WireDraggable' })
          .trigger('contextmenu')

        expect(store.setConnectionSelectionState).not.toHaveBeenCalled()
      })

      it('should emit the `contextmenu` event', async () => {
        await wrapper
          .findComponent({ name: 'WireDraggable' })
          .trigger('contextmenu')

        expect(wrapper.emitted('contextmenu')).toBeTruthy()
      })
    })

    describe('when a control point context menu is requested', () => {
      beforeEach(() => {
        stubAll(store, [
          'setControlPointSelectionState'
        ])

        store.connections[connectionId].controlPoints = [
          createControlPoint(),
          createControlPoint()
        ]
      })

      it('should select the connection if it is not already selected', async () => {
        store.connections[connectionId].controlPoints[0].isSelected = false

        await wrapper
          .findComponent({ name: 'Draggable' })
          .trigger('contextmenu')

        expect(store.setControlPointSelectionState).toHaveBeenCalledTimes(1)
        expect(store.setControlPointSelectionState).toHaveBeenCalledWith(connectionId, 0, true)
      })

      it('should not select the connection if it is already selected', async () => {
        store.connections[connectionId].controlPoints[0].isSelected = true

        await wrapper
          .findComponent({ name: 'Draggable' })
          .trigger('contextmenu')

        expect(store.setControlPointSelectionState).not.toHaveBeenCalled()
      })

      it('should emit the `contextmenu` event', async () => {
        await wrapper
          .findComponent({ name: 'Draggable' })
          .trigger('contextmenu')

        expect(wrapper.emitted('contextmenu')).toBeTruthy()
      })
    })
  })

  describe('when a wire is selected', () => {
    beforeEach(() => {
      stubAll(store, [
        'deselectAll',
        'setConnectionSelectionState'
      ])
    })

    it('should deselect all items before selecting the control point', async () => {
      await wrapper
        .findComponent({ name: 'WireDraggable' })
        .vm
        .$emit('select', true)

      expect(store.deselectAll).toHaveBeenCalledTimes(1)
      expect(store.setConnectionSelectionState).toHaveBeenCalledTimes(1)
      expect(store.setConnectionSelectionState).toHaveBeenCalledWith(connectionId, true)
    })

    it('should not deselect any items before selecting the control point', async () => {
      await wrapper
        .findComponent({ name: 'WireDraggable' })
        .vm
        .$emit('select', false)

      expect(store.deselectAll).not.toHaveBeenCalled()
      expect(store.setConnectionSelectionState).toHaveBeenCalledTimes(1)
      expect(store.setConnectionSelectionState).toHaveBeenCalledWith(connectionId, true)
    })
  })

  describe('when a control point is selected', () => {
    beforeEach(() => {
      stubAll(store, [
        'deselectAll',
        'setControlPointSelectionState'
      ])

      store.connections[connectionId].controlPoints = [
        createControlPoint(),
        createControlPoint()
      ]
    })
    it('should deselect all items before selecting the control point', async () => {
      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('select', true)

      expect(store.deselectAll).toHaveBeenCalledTimes(1)
      expect(store.setControlPointSelectionState).toHaveBeenCalledTimes(1)
      expect(store.setControlPointSelectionState).toHaveBeenCalledWith(connectionId, 0, true)
    })

    it('should not deselect any items before selecting the control point', async () => {
      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('select', false)

      expect(store.deselectAll).not.toHaveBeenCalled()
      expect(store.setControlPointSelectionState).toHaveBeenCalledTimes(1)
      expect(store.setControlPointSelectionState).toHaveBeenCalledWith(connectionId, 0, true)
    })
  })
})
