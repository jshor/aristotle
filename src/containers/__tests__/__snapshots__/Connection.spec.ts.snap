// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item container should match the snapshot 1`] = `
"<!-- the tab-accessible centroid element -->
<div style="position: absolute; left: 0px; top: 0px;" tabindex="0" aria-label="Connection from Source Port to Target Port" title="Connection from Source Port to Target Port" data-test="centroid"></div>
<!-- each segment of the visible wire -->
<!-- TODO: set is-animated according to preferences -->
<svg class="wire wire--selected" width="15" height="15" aria-roledescription="connection" style="left: -7.5px; top: -7.5px;" title="Connection from Source Port to Target Port">
  <path class="wire__outline" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
  <path class="wire__display" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
  <!--v-if-->
  <path class="wire__clickable" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
</svg>
<!-- the visual outline of the bounding box surrounding the connection -->
<!--v-if-->
<!-- the control points -->
<!-- render only n-1 control points, since the last one is the target port -->"
`;

exports[`Item container when dragging a wire snap 1`] = `
"<!-- the tab-accessible centroid element -->
<div style="position: absolute; left: 0px; top: 0px;" tabindex="0" aria-label="Connection from Source Port to Target Port" title="Connection from Source Port to Target Port" data-test="centroid"></div>
<!-- each segment of the visible wire -->
<!-- TODO: set is-animated according to preferences -->
<svg class="wire wire--selected" width="15" height="15" aria-roledescription="connection" style="left: -7.5px; top: -7.5px;" title="Connection from Source Port to Target Port">
  <path class="wire__outline" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
  <path class="wire__display" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
  <!--v-if-->
  <path class="wire__clickable" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
</svg>
<!-- the visual outline of the bounding box surrounding the connection -->
<!--v-if-->
<!-- the control points -->
<!-- render only n-1 control points, since the last one is the target port -->"
`;
